include/irr/asset/CTexturePacker.h

```diff
@@ -81,57 +80,32 @@ class ITexturePacker
         }
         uint32_t x() const { return addr&X_MASK; }
         uint32_t y() const { return (addr>>(ADDR_LAYER_SHIFT>>1))&X_MASK; }
-        uint32_t layer() const { return addr>>ADDR_LAYER_SHIFT; }
+        uint32_t layer() const { return (addr>>ADDR_LAYER_SHIFT)&0xffffu; }
+
+        bool isValid() const { return (addr&0xffffu)==invalid_addr; }
+        bool hasMipTailAddr() const { return addr&(0xffffu<<16); }
+        STexOffset getMipTailAddr() const { return addr>>16; }
 
+        //upper 16 bits are used for address of mip-tail page
         uint32_t addr;
     };
 
+    //TODO will work only with _tilesPerDim=32, for different values ADDR_LAYER_SHIFT must be adjusted as well to correctly retrieve 2d+layer addr from morton+layer
     ITexturePacker(uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u) :
         m_pgSzxy(_pgSzxy),
         m_tilesPerDim(_tilesPerDim),
-        m_addrAlctr(14u, _tilesPerDim, _numLayers)
+        m_addrAlctr(ADDR_LAYER_SHIFT>>1, _tilesPerDim, _numLayers)
     {
-        ICPUImage::SCreationParams params;
-        params.arrayLayers = _numLayers;
-        params.extent = {_tilesPerDim,_tilesPerDim,1u};
-        params.flags = static_cast<ICPUImage::E_CREATE_FLAGS>(0);
-        params.format = EF_R32_UINT;
-        params.mipLevels = core::findMSB(_tilesPerDim)+1;
-        params.samples = ICPUImage::ESCF_1_BIT;
-        params.type = IImage::ET_2D;
-        m_pageTable = ICPUImage::create(std::move(params));
-        
-        auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(m_pageTable->getCreationParameters().mipLevels);
-
-        uint32_t bufOffset = 0u;
-        for (uint32_t i = 0u; i < m_pageTable->getCreationParameters().mipLevels; ++i)
-        {
-            const uint32_t tilesPerLodDim = _tilesPerDim>>i;
-            const uint32_t regionSz = tilesPerLodDim*tilesPerLodDim*_numLayers*4u;
-            auto& region = (*regions)[i];
-            region.bufferOffset = bufOffset;
-            region.bufferImageHeight = 0u;
-            region.bufferRowLength = tilesPerLodDim*4u;
-            region.imageExtent = {tilesPerLodDim,tilesPerLodDim,1u};
-            region.imageOffset = {0,0,0};
-            region.imageSubresource.baseArrayLayer = 0u;
-            region.imageSubresource.layerCount = _numLayers;
-            region.imageSubresource.mipLevel = i;
-            region.imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
-
-            bufOffset += regionSz;
-        }
-        auto buf = core::make_smart_refctd_ptr<ICPUBuffer>(bufOffset);
-
-        m_pageTable->setBufferAndRegions(std::move(buf), regions);
+        assert(core::isPoT(_tilesPerDim));
```

Comment:
only need to be multiple (different in each dimension) of max-allocatable texture size (which itself needs to be PoT and square)

Reply:
well, i think it has to be PoT for allocation of squares with morton codes to work fine

Reply:
ah shit you're kind of right

for morton code it needs to be either

`xDim(1<<tilesPerDimLog2), yDim(1<<tilesPerDimLog2)`
or
`xDim(1<<(tilesPerDimLog2+1)), yDim(1<<tilesPerDimLog2)`

Reply:
> `xDim(1<<tilesPerDimLog2), yDim(1<<tilesPerDimLog2)`

how it is now since `1<<tilesPerDimLog2` is `tilesPerDim`

> `xDim(1<<(tilesPerDimLog2+1)), yDim(1<<tilesPerDimLog2)`

it's not even square

Reply:
> it's not even square

Yes morton codes work with textures that have an aspect ratio of 1:1 or 2:1

Reply:
Because its  Z-order and the Z covers the X first so its possible to have a 2:1 ratio domain.

In 3D Morton code works with 2:1:1, 2:2:1 and 1:1:1